content/ADR-138-social-service-authentication.md

```diff
@@ -0,0 +1,101 @@
+section of the RFC) Hint: don’t be afraid of posting illustrations! The
+level of detail here has to be enough to give the reader  a clear
+understanding of the solution - it is up to the writer to decide.
+Further detail can be addressed to satisfy comments and increase clarity.
+-->
+
+The logging in system for the social needs to solve three problems:
+
+1. The user is correctly authenticated (the token is valid)
+1. Who the user is (which userId does the token belong to)
+1. Does the user have the correct permissions to perform a certain action (Authorization)
+
+The first two problems are already solved by the current chat solution (Matrix), so we propose to leverage the already existing implementation for simplicity and velocity in development. Matrix will be used as an authentication system, users will login against their service, then each endpoint will first validate against Matrix that the user is still using a valid token.
+
+The service will have a distributed cache (without persistance) with user-to-token information which will be used both to obtain the user (once the token is validated) and to prevent hitting Matrix for every request.
+
+The login flow will be the following:
+
+```mermaid
+sequenceDiagram
+    User->>+Social service: Login
+    Social service->>+Matrix: Login
+    Matrix->>-Social service: Token
+    Social service->>+Cache: Store token
+    Social service-->>-User: Token
+```
+
+Then each subsequent request will perform the following validation:
+
+```mermaid
+graph LR
+    User -->|Request| Social-service(Social service)
+    Social-service -->|Get User | Cache
+    Cache -->|User| Social-service
+    Social-service --> C{User present}
+
+    C -->|True| D(Continue request)
+    C -->|False| Validate-token(Validate token against matrix)
+    Validate-token --> |Get User| Matrix
+    Matrix --> |User| Social-service
+```
+
+In this second workflow the token life in the cache will be extended by what's defined in matrix, thus preventing the user from using an invalid token.
+If the token is expired and the user tries to use it again, the service will validate against Matrix, if it's still valid, the user will be obtained again and re-stored in the Cache.
+
+To validate the token against Matrix, an authenticated endpoint must be chosen. The endpoint needs to satisfy the following specifications:
+
+- No to low impact on the database
+- No to low usage from the client (to prevent overloading it even more)
+- Ideally tackled by a specific worker so in the case it's necessary it would be easier to scale up
+
+The endpoint that we have chosen is `_matrix/client/r0/pushrules/` but could change if we find one that suits better.
+
+## Benefit
+
+This solution has two main benefits:
+
+1. There's no need to create a new authorization/authentication service nor system
+1. The solution is completely seamless for the user and doesn't imply no persistance of tokens from the social service side
+
+## Competition
```

Comment:
prior art ADR-44

Reply:
That would be a new authentication system, until we have that, we can continue leveraging Matrix's one

Reply:
We authenticate the chat using ethereum signatures nowadays. That is what we send to matrix. Did you consider using ethereum signatures to authenticate against this service?

Reply:
The signature is sent as login system it replaces the password, but each request is authenticated with a JWT built by synapse

Reply:
In this RFC I don't imply changing the logging in system, nor the authentication model, only how we handle the token in the social service

Reply:
I think that the fact of "handling tokens" already assumes many things of the authentication model, those are worth discussing before moving on

Reply:
Oh, the new abstract is SO good. I think we can resolve this